tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,753 @@
+	// HeadChunkFileHeaderSize is the total size of the header for the head chunk file.
+	HeadChunkFileHeaderSize = SegmentHeaderSize + 2*MintMaxtSize
+	// HeaderMintOffset is the offset where the first byte of MinT for head chunk file exists.
+	HeaderMintOffset = SegmentHeaderSize
+	// HeaderMaxtOffset is the offset where the first byte of MaxT for head chunk file exists.
+	HeaderMaxtOffset = HeaderMintOffset + MintMaxtSize
+	// MaxHeadChunkFileSize is the max size of a head chunk file.
+	// Setting size to the max int32 as setting it to max int64 crashes 64 systems too.
+	MaxHeadChunkFileSize = math.MaxInt32
+	// CRCSize is the size of crc32 sum on disk.
+	CRCSize = 4
+	// MaxHeadChunkMetaSize is the max size of an mmapped chunks minus the chunks data.
+	// Max because the uvarint size can be smaller.
+	MaxHeadChunkMetaSize = SeriesRefSize + 2*MintMaxtSize + ChunksFormatVersionSize + MaxChunkLengthFieldSize + CRCSize
+)
+
+// NewChunkDiskMapper returns a new writer against the given directory
+// using the default head chunk file duration.
+func NewChunkDiskMapper(dir string, pool chunkenc.Pool) (*ChunkDiskMapper, error) {
+	return newChunkDiskMapper(dir, DefaultHeadChunkFileMaxTimeRange, pool)
+}
+
+func newChunkDiskMapper(dir string, maxFileDuration int64, pool chunkenc.Pool) (*ChunkDiskMapper, error) {
+	if maxFileDuration <= 0 {
+		maxFileDuration = DefaultHeadChunkFileMaxTimeRange
+	}
+
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, err
+	}
+	dirFile, err := fileutil.OpenDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &ChunkDiskMapper{
+		dir:       dirFile,
+		maxFileMs: maxFileDuration,
+		pool:      pool,
+		crc32:     newCRC32(),
+	}
+
+	for i := 0; i < inBufferShards; i++ {
+		m.inBufferChunks[i] = make(map[uint64]chunkenc.Chunk)
+	}
+
+	return m, m.openMMapFiles()
+}
+
+func (cdm *ChunkDiskMapper) openMMapFiles() (returnErr error) {
+	bs := map[int]ByteSlice{}
+	cs := map[int]io.Closer{}
+	defer func() {
+		if returnErr != nil {
+			var merr tsdb_errors.MultiError
+			merr.Add(returnErr)
+			merr.Add(closeAllFromMap(cs))
+			returnErr = merr.Err()
+		}
+	}()
+
+	files, err := sequenceFilesMap(cdm.dir.Name())
+	if err != nil {
+		return err
+	}
+	if cdm.pool == nil {
+		cdm.pool = chunkenc.NewPool()
+	}
+
+	seqs := make([]int, 0, len(files))
+	for seq, fn := range files {
+		f, err := fileutil.OpenMmapFile(fn)
+		if err != nil {
+			return errors.Wrap(err, "mmap files")
+		}
+		cs[seq] = f
+		bs[seq] = realByteSlice(f.Bytes())
+		seqs = append(seqs, seq)
+	}
+
+	cdm.mmappedChunkFiles = bs
+	cdm.closers = cs
+	cdm.size = 0
+
+	// Check for unsequential files.
+	sort.Ints(seqs)
+	if len(seqs) == 0 {
+		return nil
+	}
+	lastSeq := seqs[0]
+	for _, seq := range seqs[1:] {
+		if seq != lastSeq+1 {
+			return errors.Errorf("found unsequential head chunk files %d and %d", lastSeq, seq)
+		}
+		lastSeq = seq
+	}
+
+	for i, b := range cdm.mmappedChunkFiles {
+		if b.Len() < HeadChunkFileHeaderSize {
```

Comment:
It will be automatic once we integrate this with Head. It works similar to how WAL repair works now. WAL replay fails in Head, and then we attempt the repair after that. I have it in place in the patent PR where after failing to replay the mmapped chunks, we attempt the repair.

Reply:
It turns out there is repair in place for errors on `IterateAllChunks`. I will update the parent PR after this PR is merged to repair errors from here too.